### ISSUE TITLE
Validation of County/Parish data

### ISSUE THREAD

**Issue by edutrul** [bugguide/bugguide#63]
There is currently no validation of county data. E.g., I can make a Frothbistrome county in Iowa if I want to.

Validate incoming data; USGS has provided definitive county data to validate against.

### Linked issues

https://isubit.atlassian.net/browse/BUGG-97 (open)
https://isubit.atlassian.net/browse/BUGG-133 (closed)

### Comments

John:

Attached updated AllCounties.txt now featuring Canadian provinces.

On Jan 12, 2018, at 12:59 PM, Sellers, Elizabeth <esellers@usgs.gov> wrote:

Hi John,

Thanks! Happy New Year to you too!  Sorry for the delay in responding. I'm slowly digging out from under my inbox pile. 

Yes! We implemented Canadian Provinces and sub-Province (e.g. divisions, counties, boroughs) polygon support in BISON towards the end of the summer. I've attached an export of our current US counties and States & Territories; US counties/states only; and Canadian counties/provinces but I've requested that our developers send me a better export of these tables that includes the FIPS codes for the US counties/states list as well (I'll send it to you as soon as I get it). Eventually we want to be able to generate and apply centroid coordinate pairs for Canadian counties in the same way that we do for US counties. But Canada doesn't use FIPS codes (that we know of). So we still haven't worked out what code we're going to be able to use as part of our procedures for generating and applying centroid coordinate pairs for Canadian counties/districts etc. (we don't generate state or province level centroid coordinate pairs). And of course, budgets will define whether or not we have the necessary resources for implementing this improvement in BISON in FY2018.

Btw - the latest BISON dataload is currently in the 'testing and review' stage.

Hope this helps.

Cheers, Liz

Elizabeth Sellers

-

Attaching a text file containing PHP code that defines FIPS codes as

```
$fips['Iowa']['Story'] = 19169;
```

I think we want to replace long named states with their $LOCATION_CODE abbreviations (e.g., IA instead of Iowa). This array should allow us to quickly validate (isset($fips[$state][$county])) as well as look up FIPS.

**Comment by kleintom** [comment]
A couple related questions:
* the discussion above mentions Canadian counties, but I'm not seeing Canadian counties in the code anywhere, is that something we could add?
* is GADM.org the current source for county names? If so it looks like we need to update our list - see pr 63.
* In pr 63 I did the simple thing of just comparing a user-supplied county name with what we have in bgmap.fips.inc (and sending them to GADM if they get it wrong) - that seems simple for us, but we're kind of putting the onus on users to spell and format counties correctly. Would it be better if we populated the county field as a dropdown after a state/province was selected? It would have to be done with ajax I suppose (or we could just provide a county dropdown that included every county from every state/province, then we wouldn't need ajax).
* Similar issues/question with the user profile fields 'default state/province' and 'default county' - the default county could become a select once the default state is chosen (or we can just rely on the validation we're adding to image submission to catch an incorrectly entered default county).

**Comment by clouseau** [comment]
Yes, we should add Canadian counties. Or Rural Municipality. Or Municipal District. It varies by province! So GADM.org makes sense for second-level subdivisions. We also need to sync that with the mapping system to see what they support. 

The current bgmap.fips.inc is about 116k, 28k compressed. Reasonable to send along with the form if we need to.

I think default user profile field for state/province and county should be dropdowns or at least validated.

**Comment by kleintom** [comment]
It looks like a javascript file to hold all of the county data (~3440 county units for Canada and US combined) is about 40k or 14k gzipped, so I'm thinking of updating the county select options using javascript each time a state/province is selected. (The only other ways I can think to do it are with ajax, which I think would be pretty frustrating if it was slow, or simply include all 3440 counties in the county dropdown all the time (no thanks).)

Before I get too far though, John do you have any thoughts on how to handle existing migrated counties that don't match any of the GADM county names?  Note it seems there are several valid reasons why that  might happen:
* the user entered a non-existent county or a county that doesn't exist in the state they selected (or they entered a county with no state, so we don't know which state to assign it to (without guessing));
* they entered a valid county but had a typo;
* they entered a valid county but it's formatted differently than the way GADM formats it (e.g. doesn't use a dash where GADM does, or doesn't use French accents where GADM does);
* they entered a county that was valid when they entered it, but now it's not valid (see e.g. the discussion at 421 on Alaska counties, where it appears some names changed and two counties were combined since we last updated our county list).

My first thought is that if their entered county (and state) matches a known county, we just present the counties for that state with their county selected (business as usual). If we don't recognize their county then we give a message telling them we don't recognize the county, and ask them to select a county (and state if needed) from the dropdown - if they don't select a new county then we just keep their old value (it might be an old name that was valid but is no longer, and they might not even know what the new valid county should be if e.g. a county was split in two and they're not sure which new county their pic was taken in! (could that really happen?)).  [From a technical point of view, this means the county field would remain a text field, not a 'List (text)' (which would require a list of valid counties) - the counties dropdown presented to the user would be a select form element created by us and its value would get copied into field_bgimage_county during validation. The user would have no way to set field_bgimage_county directly.]

[Another technical issue: it seems Drupal really doesn't like it when you submit a select value that didn't exist when you created the select form element (as we would be doing with counties, since we don't want to populate the initial select with 3440 counties). I guess we'd need to reset the county select value with javascript on submit and at that point assign the selected county to some hidden text input instead or something like that. :-/ ]

Related question: do we need to handle the case where javascript is turned off?

**Comment by mbaumfalk** [comment]
The lists of counties on GADM.org seem to be around 20 years out of date. When comparing to [the list of changes to counties from the Census Bureau](https://www.census.gov/programs-surveys/geography/technical-documentation/county-changes.html), I found that it lists Clifton Forge city, Virginia, which has not existed as a county equivalent since 2001.
For the US, the Census Bureau provides the most up-to-date county data that I know of, either as [tables](https://tigerweb.geo.census.gov/tigerwebmain/TIGERweb_counties_current.html) or in the form of a [shapefile](https://www2.census.gov/geo/tiger/TIGER2021/COUNTY/).

If we were to extend this to Canada, we would need a clear definition of what we would consider a "county", since that seems to be a little different in each province. Some provinces might have multiple possible definitions (Quebec and others), and others might not have any (Yukon and Newfoundland and Labrador).
Statistics Canada has a [couple of different ways](https://www150.statcan.gc.ca/n1/pub/92-160-g/92-160-g2021002-eng.htm#a5) of splitting provinces into separate areas, though these vary per province.
